### PR TITLE
Words models, controllers, views

### PR DESCRIPTION
**Pull request by avminon**
app/Word.php - added Model for words
app/Http/Controllers/CategoryController.php - removed required for the image upon edit so that user may retain old image
app/Http/Controllers/WordController.php - added controller for words
app/Http/routes.php - added Route::resource('words', 'WordController');
config/paths.php - updated paths for sound files
view/app.blade - added menu entry for Words
views/words - added views for words

### CODE REVIEW

**Comment by avminon** [comment]
image
image
image

File: app/Category.php [hunk 1]
```diff
@@ -44,4 +44,9 @@ public function assign($values)
         $this->save();
     }
 
+    public function getWords()
```
Comment (nguyenthanhtung88) [inline app/Category.php]: 1. This function name doesn't make sense, i saw you return count so this function name should be `getCountWords` for example.
2. Logic in this function is same as relationship with Word (https://github.com/framgia/fels_107/pull/13/files#diff-484994944c689c504ddbf686207c6bd1R20) => You can use relationship for listing or counting

File: app/Http/Controllers/WordController.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+class WordController extends Controller
+{
+    /**
+     * Create a new controller instance.
+     *
+     * @return void
+     */
+    public function __construct()
+    {
+        $this->middleware('auth');
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: defined in route, why you have to define again here?

File: app/Http/Controllers/WordController.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        $this->user = auth()->user();
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: defined in base `Controller`, why you define again?

File: app/Http/Controllers/WordController.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    public function index()
+    {
+        $words = Word::with('category')->paginate(20);
+        return view('words.index', ['words' => $words,
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: make `'words' => $words,` in a new line, it's prettier

File: app/Http/Controllers/WordController.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        return redirect('/words');
+    }
+
+    public function edit($id)
+    {
+        try {
+            $word = Word::findOrFail($id);
+            $categories = Category::lists('name', 'id');
+            $optionArray = explode(";", $word->options);
+            $option1 = $optionArray[0];
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: Don't use fixed variable like `$optionX` (X = 1..3), I think you should pass `$optionArray` in view and use foreach
=> This is good for future you have to add more option (not only 3) or you will solve case of different words can have differenent number of options

File: app/Http/Controllers/WordController.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                'option3' => 'required',
+                'sound' => 'mimes:wav,mp3,mp4,mpeg,mpg',
+            ]);
+            $word = Word::findOrFail($id);
+            $word->assign($request);
+            \Session::flash('flash_success', 'Update successful!');
+        } catch (ModelNotFoundException $e) {
+            \Session::flash('flash_error', 'Update failed. The word cannot be found.');
+        }
+        return redirect('/words');
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: All code snippets of this function do nearly the same like `store` function, why didn't you make it reuseable? Write a protected function of current controller is one solution.

File: app/Word.php [hunk 7]
```diff
@@ -20,4 +20,28 @@ public function lessonWords()
...
     }
 
+    public function getSoundAttribute($values)
+    {
+        return (!empty($values)) ? $values : null;
+    }
+
+    public function assign($values)
+    {
+        if ($values->input('word_id') != null) {
```
Comment (nguyenthanhtung88) [inline app/Word.php]: Please use `!is_null()` or `!empty()`

File: config/paths.php [hunk 8]
```diff
@@ -1,6 +1,8 @@
 <?php
 
 return [
-        'category_path' => base_path() . '/public/images/categories/',
-        'category_image' => 'images/categories/',
+    'category_path' => base_path() . '/public/images/categories/',
```
Comment (nguyenthanhtung88) [inline config/paths.php]: `base_path('public/images/categories')`

File: config/paths.php [hunk 9]
```diff
@@ -1,6 +1,8 @@
...
+    'category_image' => 'images/categories/',
+    'word_path' => base_path() . '/public/audio/words/',
```
Comment (nguyenthanhtung88) [inline config/paths.php]: `'word_path' => base_path('public/audio/words')`

File: resources/views/words/index.blade.php [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                <tr>
+                    <td>Category: <b>{{ $word->category->name }}</b></td>
+                </tr>
+                <tr>
+                    <td>{{ $word->word }} : {{ $word->meaning }}</td>
+                </tr>
+                <tr>
+                    <td>
+                        <audio controls>
+                            <source src="{{ url(config()->get('paths.word_sound') . $word->sound)}}" type="audio/mpeg">
```
Comment (nguyenthanhtung88) [inline resources/views/words/index.blade.php]: `$word->sound)}}` => 1 space between `)` and `}}`

File: resources/views/words/edit.blade.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                        ['required' => 'required',
+                        'placeholder' => 'Option 3',
+                        'class' => 'form-control'])
+                    !!}
+                </td>
+            </tr>
+            <tr>
+                <td colspan='2'>
+                    <audio controls>
+                        <source src="{{ url('audio/words/' . $word->sound)}}" type="audio/mpeg">
```
Comment (nguyenthanhtung88) [inline resources/views/words/edit.blade.php]: `$word->sound)}}` => 1 space between `)` and `}}`

File: resources/views/categories/index.blade.php [hunk 12]
```diff
@@ -21,7 +21,7 @@
                         </p>
                     </td>
                     <td>
-                         <p><h3><u>{{ $category->name }}</u></h3></p>
+                         <p><h3><u>{{ $category->name }}</u></h3> (No. of Words added to this category: {{ $category->getWords($category->id)}})</p>
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: 1 space between `)` and `}}`

File: resources/views/categories/index.blade.php [hunk 13]
```diff
@@ -21,7 +21,7 @@
                         </p>
                     </td>
                     <td>
-                         <p><h3><u>{{ $category->name }}</u></h3></p>
+                         <p><h3><u>{{ $category->name }}</u></h3> (No. of Words added to this category: {{ $category->getCountWords($category->id) }})</p>
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: https://github.com/framgia/fels_107/pull/13/files#diff-484994944c689c504ddbf686207c6bd1R47
=> We don't have any parameters pass to `getCountWords()` function, why you pass `$category->id` here?

**Comment by nguyenthanhtung88** [comment]
OK

**Pull request merged**